crates/compiler/unify/src/unify.rs

```diff
@@ -104,6 +104,25 @@ bitflags! {
         /// specialization algorithm is running. This has implications for the unification of
         /// unspecialized lambda sets; see [`unify_unspecialized_lambdas`].
         const LAMBDA_SET_SPECIALIZATION = Mode::EQ.bits | (1 << 2);
+        /// Like [`Mode::EQ`], but permits extension of tag union types by other tag unions that
+        /// are uninhabited. This is relevant for the unification strategy of branch and condition
+        /// types, where we would like
+        ///
+        /// x : Result Str []
+        /// when x is
+        ///     Ok x -> x
+        ///
+        /// to typecheck, because [Ok Str][] = [Ok Str, Result []][] (up to isomorphism).
+        ///
+        /// Traditionally, such types do not unify because the right-hand side is seen to have a
+        /// `Result` constructor, but for branches, that constructor is not material.
+        ///
+        /// Note that such a unification is sound, precisely because `Result` is uninhabited!
+        ///
+        /// NOTE: when run in this mode, it is expected that
+        ///   - the BRANCH TYPE is passed on the LEFT
+        ///   - the CONDITION TYPE is passed on the RIGHT
+        const EQ_WITH_EXTENSION_BY_UNINHABITED_TYPES = Mode::EQ.bits | (1 << 3);
```

Comment:
This might seem like a hack, but I think it is very reasonable. Our anonymous structural tag unions don't play well with a few things, including

- marking a tag union as closed if there is no catch-all branch in a `when` expression
- expansion by inhabited types

However, syntactically, we know that it's safe to resolve both of these limitations once the branch pattern type is known, and we are about to compare with a condition type. Hence this mode needs to be applied only in that situation.